### PR TITLE
enable parallel append

### PR DESCRIPTION
**Pull request by hsyuan**
This change introduces experimental support for parallel execution of child operators under the `append` operator by inserting redistribute motion above them.

### CODE REVIEW

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 1]
```diff
@@ -691,7 +693,7 @@ CCostModelGPDB::CostUnionAll
 {
 	GPOS_ASSERT(NULL != pcmgpdb);
 	GPOS_ASSERT(NULL != pci);
-	GPOS_ASSERT(COperator::EopPhysicalUnionAll == exprhdl.Pop()->Eopid());
+	GPOS_ASSERT(NULL != CPhysicalUnionAll::PopConvert(exprhdl.Pop()));
```
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDB.cpp]: hmm.. shouldn't now assert that it is either serial or parallel append?

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp [hunk 2]
```diff
@@ -53,7 +53,7 @@ const CCostModelGPDBLegacy::SCostMapping CCostModelGPDBLegacy::m_rgcm[] =
 
 	{COperator::EopPhysicalTVF, CostTVF},
 
-	{COperator::EopPhysicalUnionAll, CostUnionAll},
+	{COperator::EopPhysicalSerialUnionAll, CostUnionAll},
```
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDBLegacy.cpp]: why only one? what happened to parallel? Do u plan to tune the cost metrics?

File: libgpdbcost/src/CCostModelGPDB.cpp (hunk 1 above)
Comment (d) [inline libgpdbcost/src/CCostModelGPDB.cpp]: No! The whole point of polymorphism is that entities (`CCostModelGPDB::CostUnionAll`) that interface with the base class (`CPhysicalUnionAll` in this case) do *not* need the intimate knowledge of which derived class it is. Apparently we didn't change the logic of costing even though we added a new subclass.

I'd argue that `Eopid` has its place, but our abuse of that check is a blatant violation of the abstraction and indicative of terrible design

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp (hunk 2 above)
Comment (d) [inline libgpdbcost/src/CCostModelGPDBLegacy.cpp]: They are both added. I think you are looking at the commit that renmaed / extracted the base class.

File: libgpdbcost/src/CCostModelGPDB.cpp (hunk 1 above)
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDB.cpp]: ok cool

File: libgpopt/include/gpopt/base/CDistributionSpecStrictHashed.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+//	Greenplum Database
+//	Copyright (C) 2016 Pivotal Software, Inc.
+
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/base/CDistributionSpecStrictHashed.h]: can we have a comment of why we need this?

File: libgpopt/include/gpopt/operators/CPhysicalParallelUnionAll.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
+//	Greenplum Database
+//	Copyright (C) 2016 Pivotal Software, Inc.
+
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CPhysicalParallelUnionAll.h]: comment please

File: libgpopt/include/gpopt/operators/CPhysicalSerialUnionAll.h [hunk 5]
```diff
@@ -0,0 +1,140 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2012 EMC Corp.
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CPhysicalSerialUnionAll.h]: update year

File: libgpopt/include/gpopt/operators/CPhysicalSerialUnionAll.h [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+//
+//	@filename:
+//		CPhysicalSerialUnionAll.h
+//
+//	@doc:
+//		Physical union all operator
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CPhysicalSerialUnionAll.h]: update comment

File: libgpopt/include/gpopt/operators/CPhysicalUnionAllFactory.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
+//	Greenplum Database
+//	Copyright (C) 2016 Pivotal Software, Inc.
+#ifndef GPOPT_CPhysicalUnionAllFactory_H
+#define GPOPT_CPhysicalUnionAllFactory_H
+
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CPhysicalUnionAllFactory.h]: may be say that we generate either a parallel or serial physical unionall

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+namespace gpopt
+{
+	CPhysicalParallelUnionAll::CPhysicalParallelUnionAll
+		(
+			IMemoryPool *pmp,
+			DrgPcr *pdrgpcrOutput,
+			DrgDrgPcr *pdrgpdrgpcrInput,
+			ULONG ulScanIdPartialIndex
+		) : CPhysicalUnionAll
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: can we a uniform indentation style. I do not care which we pick. Just consistent

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
+//	Greenplum Database
+//	Copyright (C) 2016 Pivotal Software, Inc.
+
+#include "gpopt/operators/CPhysicalParallelUnionAll.h"
+#include "gpopt/base/CDistributionSpecStrictHashed.h"
+#include "gpopt/base/CUtils.h"
+#include "gpopt/operators/CExpressionHandle.h"
+
+namespace gpopt
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: can we just say **using namespace gpopt;** and not need { }?

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	const CHAR *CPhysicalParallelUnionAll::SzId() const
+	{
+		return "CPhysicalParallelUnionAll";
+	}
+
+	CDistributionSpec *CPhysicalParallelUnionAll::PdsDerive(IMemoryPool*, CExpressionHandle&) const
+	{
+		CDistributionSpec *pdsFirstChild = (*m_pdrgpds)[0];
+		pdsFirstChild->AddRef();
+		return pdsFirstChild;
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: Isn't this the same for CPhysicalSerialUnionAll?

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+			CDistributionSpec *,
+			ULONG ulChildIndex,
+			DrgPdp *,
+			ULONG
+		)
+	const
+	{
+		CDistributionSpec *pdsChild = (*m_pdrgpds)[ulChildIndex];
+		pdsChild->AddRef();
+		return pdsChild;
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: same here

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+				DrgPcr *pdrgpcr = (*PdrgpdrgpcrInput())[ulChild];
+				DrgPexpr *pdrgpexpr = GPOS_NEW(pmp) DrgPexpr(pmp);
+				for (ULONG ulCol = 0; ulCol < ulCols; ulCol++)
+				{
+					CExpression *pexpr = CUtils::PexprScalarIdent(pmp, (*pdrgpcr)[ulCol]);
+					pdrgpexpr->Append(pexpr);
+				}
+
+				// create a hashed distribution on input columns of the current child
+				CDistributionSpecHashed *pdshashed = GPOS_NEW(pmp) CDistributionSpecStrictHashed(pdrgpexpr, true /*fNullsColocated*/);
```
Comment (vraghavan78) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: The only difference between this and CPhysicalSerialUnionAll::BuildHashedDistributions is the distribution spec type (Strict vs. Non-Strict). So why no have a util function in CPhysicalUnionAll?

**Comment by vraghavan78** [comment]
LGTM. Great job guys!

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp (hunk 10 above)
Comment (d) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: No it's not.

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp (hunk 11 above)
Comment (d) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: No it's not

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp (hunk 12 above)
Comment (d) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: Yes we should dry it up, possibly in a template. The last thing we should do is to let the distinction in the derived subclasses leak into the base class though.

File: libgpopt/src/operators/CPhysicalParallelUnionAll.cpp (hunk 8 above)
Comment (d) [inline libgpopt/src/operators/CPhysicalParallelUnionAll.cpp]: Yes it's useless, let's kill those fillers with fire

File: libgpdbcost/src/CCostModelGPDBLegacy.cpp (hunk 2 above)
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDBLegacy.cpp]: can we remove the CCostModelGPDBLegacy code path

**Comment by vraghavan78** [comment]
LGTM!

**Comment by d** [comment]
New push addressing all comments, @vraghavan78 wanna take a look?

Thanks,
@lpetrov-pivotal and @d

**Comment by wfnuser** [comment]
I'm recently playing with GPORCA and trying to bring some new feature to it. And I found this pr is a good study case for beginners. I think after pg introduced parallel append, maybe we can enable the ORCA support for it?

And I'm a little confused with why parallel append need a redistribute motion beneath it. If anyone can explain it for me, it will be really nice.

**Pull request merged**